### PR TITLE
Many fixes and additions to support QA-310.

### PR DESCRIPTION
**Pull request by kacf**
https://tracker.mender.io/browse/QA-310

### CODE REVIEW

**Comment by kacf** [comment]
Updated! I had to rewrite the test a little bit to work on Gitlab.

The `TEST_RELEASE_TOOL_LIST_OPEN_SOURCE_ONLY` variable is a bit evil. However I couldn't find a good way without significantly more effort, because the `release_tool.py --list` command is invoked several times as sub-commands, so I can't just pass in an option. Let me know if you have any better ideas than what I proposed.

**Comment by kacf** [comment]
@mender-test-bot start pipeline

File: extra/statistics-generator [hunk 1]
```diff
@@ -112,14 +112,34 @@ collect_changes() {
 
 post_process() {
     sed -nre '
+:start
+# If empty line.
```
Comment (oleorhagen) [inline extra/statistics-generator:116]: SO MUCH YES!

File: extra/release_tool.py [hunk 2]
```diff
@@ -2460,17 +2460,30 @@ def log_and_run_cmd(cmd, workdir):
...
 
 
-def do_generate_release_notes(state):
-    """Generate Release Notes for the released tags"""
+def do_generate_release_notes(
+    base_dir, version_of_integration, prev_of_integration=None
+):
+    split = version_of_integration.split("..")
+    if len(split) > 1:
+        if prev_of_integration is not None:
```
Comment (oleorhagen) [inline extra/release_tool.py:2484]: Suggestion:
```
if prev_of_integration is not None
```
=
```
if prev_of_integration
```

File: extra/release_tool.py [hunk 3]
```diff
@@ -2460,17 +2460,30 @@ def log_and_run_cmd(cmd, workdir):
         fd.write(changelog_out.decode())
 
 
-def do_generate_release_notes(state):
-    """Generate Release Notes for the released tags"""
```
Comment (oleorhagen) [inline extra/release_tool.py:2612]: Q:
Why remove the comment all together?

File: extra/release_tool.py [hunk 4]
```diff
@@ -2460,17 +2460,30 @@ def log_and_run_cmd(cmd, workdir):
...
+def do_generate_release_notes(
+    base_dir, version_of_integration, prev_of_integration=None
+):
+    split = version_of_integration.split("..")
+    if len(split) > 1:
+        if prev_of_integration is not None:
+            raise Exception(
+                "version_of_integration is a range, and prev_of_integration is non-empty. This should not happen!"
+            )
+        prev_of_integration = split[0]
```
Comment (oleorhagen) [inline extra/release_tool.py:2488]: Tricks:
`prev_of_integration, version_of_integration = split[0], split[1]`
But splitting hairs ofc :)

File: extra/release_tool.py [hunk 5]
```diff
@@ -2460,17 +2460,30 @@ def log_and_run_cmd(cmd, workdir):
...
+        version_of_integration = split[1]
+
+    Component.set_integration_version(version_of_integration)
+
+    if prev_of_integration is None:
```
Comment (oleorhagen) [inline extra/release_tool.py:2493]: Same as above.

File: extra/release_tool.py [hunk 6]
```diff
@@ -3087,10 +3107,10 @@ def main():
         "--in-integration-version",
         dest="in_integration_version",
         metavar="VERSION",
-        help="Used together with the above argument to query for a version of a "
+        help="Used together with the `-g` argument to query for a version of a "
```
Comment (oleorhagen) [inline extra/release_tool.py:3126]: Suggestion:
Use the long option in the help text?
:)

File: extra/release_tool.py [hunk 7]
```diff
@@ -3087,10 +3107,10 @@ def main():
...
         + "service which is in the given version of integration, instead of the "
         + "currently checked out version of integration. If a range is given here "
-        + "it will return the range of the corresponding service.",
+        + "it will return the range of the corresponding service. It is also used together with the --generate-release-notes argument.",
```
Comment (oleorhagen) [inline extra/release_tool.py:3129]: Nit:
quote --generate-releasenotes endquote ?

File: extra/statistics-generator [hunk 8]
```diff
@@ -63,7 +63,7 @@ collect_changes() {
     else
         # Sorting is important because we need "*-enterprise" repositories to
         # come after their Open Source counterparts (see below).
-        REPO_LIST="$("$RELEASE_TOOL" --list git --only-backend | sort)"
+        REPO_LIST="$("$RELEASE_TOOL" --list git --in-integration-version $1 --only-backend | sort)"
         REPO_LIST="$REPO_LIST integration"
```
Comment (oleorhagen) [inline extra/statistics-generator:67]: Question:
In newer branches `integration` will be present twice now, is this a problem?

**Review by oleorhagen — APPROVED** [review]
Oh my lord! 

What a pleasure!

File: extra/release_tool.py (hunk 3 above)
Comment (kacf) [reply extra/release_tool.py:2612]: Brainfart. I have no idea why I removed it, I will put it back.

File: extra/release_tool.py (hunk 2 above)
Comment (kacf) [reply extra/release_tool.py:2484]: Strictly speaking it doesn't matter in this case, but I would like it to match the default argument, and the lower variant will match `""` and `0` as well.

File: extra/statistics-generator (hunk 8 above)
Comment (kacf) [reply extra/statistics-generator:67]: `release_tool` doesn't list itself (integration), so this is correct, actually. But you're right that this is a gotcha, I'm not sure if I've taken this into account every time I've written a piece of code which loops over this.

File: extra/release_tool.py (hunk 6 above)
Comment (kacf) [reply extra/release_tool.py]: It's pre-existing, but worth a cleanup anyway, enjoy https://github.com/mendersoftware/integration/pull/1836/commits/95ad4e735d036e20bbb88d8afe00bfc2841ef263!

**Comment by kacf** [comment]
@mender-test-bot start pipeline

File: extra/release_tool.py [hunk 9]
```diff
@@ -263,6 +272,9 @@ def is_independent_component(self):
 BACKEND_SERVICES = (
     BACKEND_SERVICES_OPEN | BACKEND_SERVICES_ENT | BACKEND_SERVICES_OPEN_ENT
 )
+CLIENT_SERVICES_ENT = {
```
Comment (lluiscampos) [inline extra/release_tool.py:275]: Is this strictly necessary? :cry: 

Update note in [point 3 here](https://github.com/mendersoftware/mender-qa/blob/master/Documentation/adding-new-repository.md#integration) so that it is followed for all repos (not only backend repos)

File: extra/test/release_notes_mender-artifact.txt [hunk 10]
```diff
@@ -0,0 +1,10 @@
+## mender-artifact 3.6.1
+
+_Released 09.28.2021_
+
+### Statistics
+
+A total of 0 lines added, 0 removed (delta 0)
+
+### Changelogs
+
```
Comment (lluiscampos) [inline extra/test/release_notes_mender-artifact.txt:10]: These (here but also mender-cli and mender-connect) do not look correct to me. There must be statistics and changelogs.

File: extra/statistics-generator (hunk 8 above)
Comment (kacf) [reply extra/statistics-generator:67]: I now came across this again, and let me make a small modification: It doesn't list itself **when using `--only-backend`**. We can debate whether this is correct I suppose, but I'm not going to change it now. If we do decide to change it later, this argument is pretty easy to search for and change everywhere.

File: extra/release_tool.py (hunk 9 above)
Comment (kacf) [reply extra/release_tool.py:275]: Thanks, I'll do that!

For whether it's necessary, see [this comment](https://github.com/mendersoftware/integration/pull/1836#issuecomment-1005715282). I welcome other approaches, this part is tricky because it is invoked so many times as a sub command.

File: extra/test/release_notes_mender-artifact.txt (hunk 10 above)
Comment (kacf) [reply extra/test/release_notes_mender-artifact.txt:10]: That depends. What you mean is that this file should be removed actually, since there are no changes. This makes sense for releases, where it's a manual process, and we just won't paste it if it says "0 lines".

But [QA-310](https://tracker.mender.io/browse/QA-310) says to "completely replace" what's there, which I take to mean that it shouldn't try to mix it in with the content which is already in mender-docs-changelog, so if nothing is generated, this component won't even exist in the menu. We could try to mix it in with the pre-existing text, but it's more error prone. What do you think?

Comment (kacf) [reply extra/test/release_notes_mender-artifact.txt:10]: It's good that you pointed it out though, because it looks like they *always* are zero, even when they shouldn't be. I'll fix that, but my point above still stands.

Comment (lluiscampos) [reply extra/test/release_notes_mender-artifact.txt:10]: Yes, what I meant is that this specific release, `mender-artifact 3.6.1` cannot be blank.

Comment (kacf) [reply extra/test/release_notes_mender-artifact.txt:10]: No, because mender-artifact in integration 3.0.1 is 3.6.1, and in integration 3.1.0, it's also 3.6.1. So the queried range is `3.6.1..3.6.1`, in other words empty. What I could do is to change the test to test integration `3.0.0..3.1.0` instead of `3.0.1..3.1.0`. This diff would then not be empty anymore, but some of the others still would be, so this might be a test with better coverage of the corner cases. I think I'll do that.

Comment (kacf) [reply extra/test/release_notes_mender-artifact.txt:34]: Done! See https://github.com/mendersoftware/integration/pull/1836/commits/0f34e8573b3937ae3ee1ed462806368b0f30e122.

Note that mender-cli is still empty, as it should be.

Comment (lluiscampos) [reply extra/test/release_notes_mender-artifact.txt:34]: I understand it now! Thanks.

Yes, that range seems to be better for test purposes.

**Review by lluiscampos — APPROVED** [review]

File: extra/test_release_tool.py [hunk 11]
```diff
@@ -142,25 +142,67 @@ def test_version_of(capsys):
...
+    # multiple containers and we don't know which one to choose.
     run_main_assert_result(capsys, ["--version-of", "mender"], "master")
+    try:
+        run_main_assert_result(
+            capsys, ["--version-of", "mender", "--version-type", "docker"], "master"
+        )
+    except:
+        pass
+    else:
+        pytest.fail("Expected exception!")
```
Comment (lluiscampos) [inline extra/test_release_tool.py:155]: An alternative using [pytest functionality](https://docs.pytest.org/en/6.2.x/reference.html#pytest.raises) would be:
```
with pytest.raises(RuntimeError):
    run_main_assert_result(
        capsys, ["--version-of", "mender", "--version-type", "docker"], "master"
    )
```

**Review by lluiscampos — COMMENTED** [review]
Thanks for the human friendly commit messages. The tests in the last commit are also very good to illustrate the changes.

I'll give it a try now pretending to do a new release :+1:

**Review by lluiscampos — APPROVED** [review]
> I'll give it a try now pretending to do a new release +1

It passed my test :heavy_check_mark:

**Comment by kacf** [comment]
@mender-test-bot start pipeline

File: extra/test_release_tool.py (hunk 11 above)
Comment (kacf) [reply extra/test_release_tool.py:155]: Thanks, another nugget to remember! Changed.

Comment (kacf) [reply extra/test_release_tool.py:155]: And by changed, I mean, I did it locally now, and will apply it together with the test re-enabling.

**Comment by kacf** [comment]
The prebuilt client failure is expected because the image has changed name. This should pass on master after this is merged and the client is rebuilt.

Merging.
